service/src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceTest.java

```diff
@@ -1086,6 +1097,275 @@ public void updateBqDatasetWithInvalidExpirationTimes() throws Exception {
+
+    String location = "us-central1";
+    ControlledGcsBucketResource resource =
+        ControlledResourceFixtures.makeDefaultControlledGcsBucketResource()
+            .workspaceId(workspace.getWorkspaceId())
+            .accessScope(AccessScopeType.ACCESS_SCOPE_SHARED)
+            .managedBy(ManagedByType.MANAGED_BY_USER)
+            .build();
+
+    // Test idempotency of bucket-specific undo steps by retrying them once. Fail at the end of
+    // the flight to ensure undo steps work properly.
+    Map<String, StepStatus> retrySteps = new HashMap<>();
+    retrySteps.put(CreateGcsBucketStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    retrySteps.put(GcsBucketCloudSyncStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    jobService.setFlightDebugInfoForTest(
+        FlightDebugInfo.newBuilder().undoStepFailures(retrySteps).lastStepFailure(true).build());
+    // JobService throws a InvalidResultStateException when a synchronous flight fails without an
+    // exception, which occurs when a flight fails via debugInfo.
+    assertThrows(
+        InvalidResultStateException.class,
+        () ->
+            controlledResourceService.createBucket(
+                resource,
+                ControlledResourceFixtures.getGoogleBucketCreationParameters(),
+                Collections.emptyList(),
+                user.getAuthenticatedRequest()));
+
+    // Validate the bucket does not exist.
+    StorageCow storageCow = crlService.createStorageCow(projectId);
+    assertNull(storageCow.get(resource.getBucketName()));
+
+    assertThrows(
+        ResourceNotFoundException.class,
+        () ->
+            controlledResourceService.getControlledResource(
+                workspace.getWorkspaceId(),
+                resource.getResourceId(),
+                user.getAuthenticatedRequest()));
+  }
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = BUFFER_SERVICE_DISABLED_ENVS_REG_EX)
+  void deleteGcsBucketDo() throws Exception {
+    UserAccessUtils.TestUser user = userAccessUtils.defaultUser();
+    Workspace workspace = reusableWorkspace(user);
+    String projectId =
+        workspaceService.getAuthorizedRequiredGcpProject(
+            workspace.getWorkspaceId(), user.getAuthenticatedRequest());
+
+    ControlledGcsBucketResource resource =
+        ControlledResourceFixtures.makeDefaultControlledGcsBucketResource()
+            .workspaceId(workspace.getWorkspaceId())
+            .accessScope(AccessScopeType.ACCESS_SCOPE_SHARED)
+            .managedBy(ManagedByType.MANAGED_BY_USER)
+            .build();
+
+    ControlledGcsBucketResource createdBucket =
+        controlledResourceService.createBucket(
+            resource,
+            ControlledResourceFixtures.getGoogleBucketCreationParameters(),
+            Collections.emptyList(),
+            user.getAuthenticatedRequest());
+    assertEquals(resource, createdBucket);
```

Comment:
I think these first 2 blocks are repeated in most methods, and the last block is repeated in some methods. Can you pull it out into a helper method, maybe that updates local properties or something?